File: scripts/move_arm_with_plan.go
package main

import (
	"context"
	"encoding/csv"
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"

	"github.com/edaniels/golog"
	pb "go.viam.com/api/component/arm/v1"
	"go.viam.com/rdk/components/arm"
	"go.viam.com/rdk/robot/client"
	rdkutils "go.viam.com/rdk/utils"
	"go.viam.com/utils"
	"go.viam.com/utils/rpc"
)

var logger = golog.NewDevelopmentLogger("client")

func main() {
	utils.ContextualMain(mainWithArgs, logger)
}

func mainWithArgs(ctx context.Context, args []string, logger golog.Logger) error {
	// parse command line input
	address := flag.String("address", "", "address of robot on app.viam.com")
	payload := flag.String("payload", "", "payload corresponding to address")
	sceneName := flag.String("scene_name", "", "name of the scene plan file to parse and execute")
	armName := flag.String("arm", "arm", "name of the arm being planned with")
	flag.Parse()

	// Connect to robot arm
	robot, err := connect(*address, *payload)
	if err != nil {
		return err
	}
	defer robot.Close(context.Background())
	fmt.Println(">>> Successfully connected to robot")
	arm, err := arm.FromRobot(robot, *armName)
	if err != nil {
		return err
	}
	fmt.Printf(">>> Found arm with name [ %s ]\n", *armName)

	// Parse the plan for the given scene name
	path := resolvePath(*sceneName + ".csv")
	fmt.Printf(">>> Searching for plan path file [ %s ]\n", path)
	plan, err := planFromCSV(path)
	if err != nil {
		return err
	}

	// Execute the plan
	fmt.Println(">>> Executing plan ...")
	for i, step := range plan {
		fmt.Printf("\t\tStep [ %d / %d ] - %.2v\n", i, len(plan), step.Values)
		err := arm.MoveToJointPositions(context.Background(), step, nil)
		if err != nil {
			return err
		}
	}
	return nil
}

func resolvePath(filename string) string {
	_, thisFilePath, _, _ := runtime.Caller(0)
	thisDirPath, err := filepath.Abs(filepath.Dir(thisFilePath))
	if err != nil {
		panic(err)
	}
	return filepath.Join(thisDirPath, "..", filename)
}

func planFromCSV(filepath string) ([]*pb.JointPositions, error) {
	// open file
	f, err := os.Open(filepath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	// read csv values using csv.Reader
	csvReader := csv.NewReader(f)
	data, err := csvReader.ReadAll()
	if err != nil {
		return nil, err
	}

	// convert to a plan
	plan := make([]*pb.JointPositions, 0)
	for _, line := range data {
		floats := make([]float64, 0)
		for _, elem := range line {
			float, err := strconv.ParseFloat(elem, 64)
			if err != nil {
				return nil, err
			}
			floats = append(floats, rdkutils.RadToDeg(float))
		}
		plan = append(plan, &pb.JointPositions{Values: floats})
	}
	return plan, nil
}

func connect(address, payload string) (*client.RobotClient, error) {
	credentials := rpc.WithCredentials(rpc.Credentials{Type: rdkutils.CredentialsTypeRobotLocationSecret, Payload: payload})
	return client.New(context.Background(), address, logger, client.WithDialOptions(credentials))
}

File: scripts/move_arm_with_plan.py
#!/usr/bin/env python
#
# Moves a robotic arm with a motion plan as defined by a given scene with the PySDK
#
# This script assumes the following: 1) that the given scene path file is accessible in the directory above, 2) that the
# number of joints on the connected robot arm matches the number of joints in the plan file, and 3) that the kinematic
# model of the connected arm matches the kinematic model used when solving the initial plan.

import argparse
import asyncio
import csv
import numpy
import os
import sys

from viam.components.arm import Arm, JointPositions
from viam.robot.client import RobotClient
from viam.rpc.dial import Credentials, DialOptions

ARM_COMPONENT_NAME = "MANIPULATOR"

async def connect():
    creds = Credentials(type='robot-location-secret', payload='<YOUR PAYLOAD HERE>')
    opts = RobotClient.Options(refresh_interval=0, dial_options=DialOptions(credentials=creds))
    return await RobotClient.at_address('<YOUR ADDRESS HERE>', opts)

async def main():
    # Get the scene whose plan we want to use
    parser = argparse.ArgumentParser(description='Executes a motion plan on the connected arm using the PySDK')
    parser.add_argument('scene_name', type=str, help='Name of the scene plan file to parse and execute')
    args = parser.parse_args()

    # Parse the plan for the given scene name
    plan_csv = args.scene_name + ".csv"
    plan_filepath = os.path.join(os.path.abspath(os.path.curdir), "..", plan_csv)
    print(">>> Searching for plan path file [ {0} ]".format(plan_filepath))
    if os.path.isfile(plan_filepath):
        print(">>> Executing plan generated for [ {0} ] with [ {1} ] : Loading [ {2} ]...".format(
              args.scene_name, ARM_COMPONENT_NAME, plan_filepath))
    else:
        print("!!! Requested scene plan file not found... Exiting.")
        sys.exit(1)

    plan = []
    with open(plan_filepath, newline='') as csvfile:
        plan_reader = csv.reader(csvfile, delimiter=',')
        for row in plan_reader:
            joints = JointPositions(values=[numpy.rad2deg(float(value)) for value in row])
            plan.append(joints)

    # Connect to the robot arm 
    robot = await connect()
    arm = Arm.from_robot(robot, ARM_COMPONENT_NAME)

    # Execute the plan
    print("??? Moving robot arm to each joint position in motion plan...")
    for target in plan:
        await arm.move_to_joint_positions(target)
    print(">>> Arm motion has completed, returning to zero position...")

    # Set a JointPositions struct full of zeroes with the same length as a state in the plan
    await arm.move_to_joint_positions(JointPositions(values=[0 for idx in range(len(plan[0].values))]))

    await robot.close()


if __name__ == '__main__':
    asyncio.run(main())
